libs/isograph-react/src/core/makeNetworkRequest.ts

```diff
@@ -438,9 +506,62 @@ function fetchNormalizationAstAndRetainArtifact<
   return retainedQuery;
 }
 
+function makeOptimisticUpdate<
+  TReadFromStore extends UnknownTReadFromStore,
+  TClientFieldValue,
+  TNormalizationAst extends NormalizationAst | NormalizationAstLoader,
+  TRawResponseType extends NetworkResponseObject,
+>(
+  environment: IsographEnvironment,
+  artifact:
+    | RefetchQueryNormalizationArtifact
+    | IsographEntrypoint<
+        TReadFromStore,
+        TClientFieldValue,
+        TNormalizationAst,
+        TRawResponseType
+      >,
+  variables: ExtractParameters<TReadFromStore>,
+  optimisticNetworkResponse: TRawResponseType,
+): OptimisticStoreLayer {
+  const root = { __link: ROOT_ID, __typename: artifact.concreteType };
+
+  if (
+    artifact.networkRequestInfo.normalizationAst.kind ===
+    'NormalizationAstLoader'
+  ) {
+    throw new Error(
+      'Using lazy loaded normalizationAst with optimisticNetworkResponse is not supported.',
+    );
+  }
+  const encounteredIds: EncounteredIds = new Map();
+  const optimistic = (environment.store =
+    addOptimisticNetworkResponseStoreLayer(environment.store));
+  normalizeData(
+    environment,
+    environment.store,
+    artifact.networkRequestInfo.normalizationAst.selections,
+    optimisticNetworkResponse,
+    variables,
+    root,
+    encounteredIds,
+  );
+
+  logMessage(environment, () => ({
+    kind: 'AfterNormalization',
```

Comment:
Hm... I think we should probably have two separate events: `WroteOptimisticUpdate` and `RevertedOptimisticUpdate`